[BUG] array out of bound in orc writer

I was trying to build libcudf with gcc-10 and it complained about this error. This looks like it might be a legitimate bug.
```console
/home/rou/src/cudf/cpp/src/io/orc/writer_impl.cu: In member function ‘void cudf::io::detail::orc::writer::impl::write_index_stream(int32_t, int32_t, cudf::io::detail::orc::orc_column_view*, size_t, size_t, size_t, size_t, const hostdevice_vector<cudf::io::orc::gpu::EncChunk>&, const hostdevice_vector<cudf::io::orc::gpu::StripeStream>&, const hostdevice_vector<cudf::io::gpu_inflate_status_s>&, cudf::io::orc::StripeInformation&, std::vector<cudf::io::orc::Stream>&, cudf::io::orc::ProtobufWriter*)’:
/home/rou/src/cudf/cpp/src/io/orc/writer_impl.cu:868:157: error: array subscript 4 is outside array bounds of ‘std::array<int, 4> [1]’ [-Werror=array-bounds]
  868 |       while ((record[1] >= 0) && (static_cast<size_t>(record[0]) >= compression_blocksize_) &&
      |                                                                                                                                                             ^                               
/home/rou/src/cudf/cpp/src/io/orc/writer_impl.cu:844:31: note: while referencing ‘present’
  844 |   std::array<int32_t, 4> present;
      |                               ^~~    
/home/rou/src/cudf/cpp/src/io/orc/writer_impl.cu:868:157: error: array subscript 4 is outside array bounds of ‘std::array<int, 4> [1]’ [-Werror=array-bounds]
  868 |       while ((record[1] >= 0) && (static_cast<size_t>(record[0]) >= compression_blocksize_) &&
      |                                                                                                                                                             ^                               
/home/rou/src/cudf/cpp/src/io/orc/writer_impl.cu:845:31: note: while referencing ‘data’
  845 |   std::array<int32_t, 4> data;
      |                               ^   
/home/rou/src/cudf/cpp/src/io/orc/writer_impl.cu:868:157: error: array subscript 4 is outside array bounds of ‘std::array<int, 4> [1]’ [-Werror=array-bounds]
  868 |       while ((record[1] >= 0) && (static_cast<size_t>(record[0]) >= compression_blocksize_) &&
      |                                                                                                                                                             ^                               
/home/rou/src/cudf/cpp/src/io/orc/writer_impl.cu:846:31: note: while referencing ‘data2’
  846 |   std::array<int32_t, 4> data2;
      |                               ^    
cc1plus: all warnings being treated as errors
CMakeFiles/cudf.dir/build.make:1056: recipe for target 'CMakeFiles/cudf.dir/src/io/orc/writer_impl.cu.o' failed
make[2]: *** [CMakeFiles/cudf.dir/src/io/orc/writer_impl.cu.o] Error 1
CMakeFiles/Makefile2:489: recipe for target 'CMakeFiles/cudf.dir/all' failed
make[1]: *** [CMakeFiles/cudf.dir/all] Error 2
Makefile:159: recipe for target 'all' failed
make: *** [all] Error 2
```

@j-ieong @harrism @jrhemstad

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -99,6 +99,7 @@
 - PR #6137 Fix issue where `np.nan` is being return instead of `NAT` for datetime/duration types
 - PR #6298 Fix gcc-9 compilation error in dictionary/remove_keys.cu
 - PR #6172 Fix slice issue with empty column
+- PR #6342 Fix array out-of-bound errors in Orc writer
 - PR #6154 Warnings on row-wise op only when non-numeric columns are found.
 - PR #6150 Fix issue related to inferring `datetime64` format with UTC timezone in string data
 - PR #6179 `make_elements` copies to `iterator` without adjusting `size`

--- a/cpp/src/io/csv/csv_common.h
+++ b/cpp/src/io/csv/csv_common.h
@@ -16,6 +16,8 @@
 
 #pragma once
 
+#include <cstdint>
+
 class SerialTrieNode;
 
 namespace cudf {

--- a/cpp/src/io/orc/writer_impl.cu
+++ b/cpp/src/io/orc/writer_impl.cu
@@ -866,9 +866,9 @@ void writer::impl::write_index_stream(int32_t stripe_id,
     if (record[0] >= 0) {
       record[0] += chunk.strm_len[type];
       while ((record[1] >= 0) && (static_cast<size_t>(record[0]) >= compression_blocksize_) &&
-             (record[3] + 3 + comp_out[record[1]].bytes_written < static_cast<size_t>(record[4]))) {
+             (record[2] + 3 + comp_out[record[1]].bytes_written < static_cast<size_t>(record[3]))) {
         record[0] -= compression_blocksize_;
-        record[3] += 3 + comp_out[record[1]].bytes_written;
+        record[2] += 3 + comp_out[record[1]].bytes_written;
         record[1] += 1;
       }
     }

--- a/cpp/src/strings/char_types/is_flags.h
+++ b/cpp/src/strings/char_types/is_flags.h
@@ -15,6 +15,8 @@
  */
 #pragma once
 
+#include <cstdint>
+
 //
 // 8-bit flag for each code-point.
 //

